component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordChangeEnforcerOnExpiration.java

```diff
@@ -234,20 +234,38 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
                     log.debug("Updated user credentials of " + tenantAwareUsername);
                 }
             } catch (org.wso2.carbon.user.core.UserStoreException e) {
-                if(e.getCause() instanceof PrivilegedActionException) {
+                if(e.getMessage().contains("InvalidOperation")){
                     if (log.isDebugEnabled()) {
-                        log.debug("Credential not valid. Credential must be a non null string with following format, "
-                                + userStoreManager.getRealmConfiguration().getUserStoreProperty("PasswordJavaRegEx"));
+                        log.debug("Invalid operation. User store is read only.", e);
                     }
-                    String errorMsg = userStoreManager.getRealmConfiguration()
-                            .getUserStoreProperty("PasswordJavaRegExViolationErrorMsg");
-                    if (errorMsg != null) {
-                        throw new AuthenticationFailedException(errorMsg);
+                    throw new AuthenticationFailedException(
+                            "Invalid operation. User store is read only", e);
+                }
+                if(e.getMessage().contains("PasswordInvalid")){
+                    if (log.isDebugEnabled()) {
+                        log.debug("Invalid credentials. Cannot proceed with the password change.", e);
+                    }
+                    throw new AuthenticationFailedException(
+                            "Invalid credentials. Cannot proceed with the password change.", e);
+                }
+                String errorMsg = userStoreManager.getRealmConfiguration()
+                        .getUserStoreProperty("PasswordJavaRegExViolationErrorMsg");
+                if(StringUtils.isNotEmpty(errorMsg) && e.getMessage().contains(errorMsg)){
+                    if (log.isDebugEnabled()) {
+                        log.debug(errorMsg, e);
+                    }
+                    throw new AuthenticationFailedException(errorMsg);
```

Comment:
we are swallowing the exception here too. It needs throw new AuthenticationFailedException(errorMsg, e)